### PR TITLE
Replace Mockito invocation counter in MetricsCompressorTest [HZ-2538]

### PR DESCRIPTION
**Pull request by JamesHazelcast**
We have recently been encountering a variety of edge-case issues with Mockito related failures in some of our tests - these tests are generally non-deterministic, but surface during long-running test jobs on Jenkins. There has been extensive investigation of these issues by multiple team members, but the issues always point towards an issue within Mockito itself (although exact causes have not been identified).

The MetricsCompressorTest encounters the same failures for the same set of tests, across multiple JVM vendors and versions. The failure is always related to the `verify(supplierSpy, times(2)).get();` check, with `times(2)` being `only()` in some tests (although still failing).

Due to the functionality of this being very simple in this test, we have decided to avoid spending more time investigating the specifics of why Mockito is struggling here and simply replace the problematic Mockito functionality with our own implementation. In this test that is a simple wrapper for the `DefaultMetricDescriptorSupplier` which uses an `AtomicInteger` to count the invocations on the `#get()` method.

Fixes https://github.com/hazelcast/hazelcast/issues/23504

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/MetricsCompressorTest.java [hunk 1]
```diff
@@ -51,49 +50,49 @@
...
 
     private final DefaultMetricDescriptorSupplier supplier = new DefaultMetricDescriptorSupplier();
-    private final Supplier<? extends MetricDescriptor> supplierSpy = spy(supplier);
+    private final SpiedMetricDescriptor spiedSupplier = new SpiedMetricDescriptor(supplier);
     private final MetricsCompressor compressor = new MetricsCompressor();
 
     @Test
     public void testSingleLongMetric() {
         MetricDescriptor originalMetric = supplier.get()
-                                                  .withPrefix("prefix")
```
Comment (TomaszGaweda) [inline hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/MetricsCompressorTest.java:64]: This is not re-formatting on my machine. Are you sure you have styles imported?

Comment (JamesHazelcast) [reply hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/MetricsCompressorTest.java]: > This is not re-formatting on my machine. Are you sure you have styles imported?

Turns out at some point I managed to reset them! Thanks for the catch, sorted now 👍

**Comment by JamesHazelcast** [comment]
run-lab-run

File: hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/MetricsCompressorTest.java [hunk 2]
```diff
@@ -263,15 +262,15 @@ public void testTwoMetrics_differentExcludedTargets() {
...
 
         verify(metricConsumerMock).consumeLong(metric1, 42L);
         verify(metricConsumerMock).consumeLong(metric2, 43L);
         verifyNoMoreInteractions(metricConsumerMock);
-        verify(supplierSpy, times(2)).get();
+        assertEquals(2, spiedSupplier.getInvocationCount());
     }
 
-    @Test
+    @Test // Passes
```
Comment (TomaszGaweda) [inline hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/MetricsCompressorTest.java:273]: unnecessary comment ;) same in test below

File: hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/MetricsCompressorTest.java [hunk 3]
```diff
@@ -51,11 +50,11 @@
 public class MetricsCompressorTest {
 
     private static final String LONG_NAME = Stream.generate(() -> "a")
-                                                 .limit(MetricsDictionary.MAX_WORD_LENGTH + 1)
-                                                 .collect(Collectors.joining());
+                                                  .limit(MetricsDictionary.MAX_WORD_LENGTH + 1)
+                                                  .collect(Collectors.joining());
```
Comment (TomaszGaweda) [inline hazelcast/src/test/java/com/hazelcast/internal/metrics/impl/MetricsCompressorTest.java:54]: please reformat again the code to revert accidental changes like here and below

**Review by TomaszGaweda — COMMENTED** [review]
Only small changes left before accept ;)
